src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,673 @@
+        dataloader (Iterable):
+            The dataloader to use for the forward pass. If None, eva_state_dict needs to be provided.
+        eva_state_dict (Optional[dict]):
+            The state_dict to load into the model. If None, a dataloader needs to be provided and the state_dict will
+            be computed using `get_eva_state_dict`.
+        forward_fn (callable):
+            The forward function to use for the forward pass. Takes two arguments: `model` and `inputs`. Default
+            behavior is `return model(**inputs)`
+        prepare_model_inputs_fn (Optional[callable]):
+            This function receives the model inputs and the peft_config and passes the output to
+            `prepare_layer_inputs_fn`. Can be used to modify the input to the SVD computation based on the original
+            model inputs. For example for language modeling the attention mask is used to determine which indices are
+            padding tokens and should not be used for SVD. Any function defined here expects two arguments:
+            `model_input` and `peft_config`. `peft.tuners.lora.eva.prepare_model_inputs_fn_language_modeling` is used
+            by default.
+        prepare_layer_inputs_fn (Union[callable, Dict[str, callable], None]):
+            This function receives the layer inputs, the model inputs (potentially modified by
+            `prepare_model_inputs_fn`) and the name of the layer and returns the inputs that should be used for SVD for
+            that particular layer. Any custom function defined here expects three arguments: `layer_input`,
+            `model_input`, and `layer_name` and should return a 2d tensor. The default logic can be found in
+            peft.tuners.lora.eva.prepare_layer_inputs_fn_language_modeling and works for language modeling. In this
+            case model_inputs is the mask used to determine which indices should be used for SVD (created by
+            `prepare_model_inputs_fn_language_modeling`).
+        adapter_name (str): The name of the adapter to initialize the weights for.
+        show_progress_bar (bool): Whether to show a progress bar. Default is True.
+
+    Returns:
+        model (torch.nn.Module): The model with the initialized LoRA weights.
+    """
+    if not hasattr(model, "peft_config"):
+        raise ValueError("model must be a PeftModel")
+
+    # eva currently only works with a single active adapter
+    # Important: when removing this requirement, make sure eva init works correctly if the new rank is 0.
+    if len(model.active_adapters) > 1:
+        raise ValueError("`initialize_lora_eva_weights` currently only works with a single active adapter")
+
+    # initialize_lora_eva_weights only works with `init_lora_weights='eva'`
+    if model.peft_config[adapter_name].init_lora_weights != "eva":
+        raise ValueError("`initialize_lora_eva_weights` can only be used with `init_lora_weights='eva'`")
+
+    # compute svd
+    if eva_state_dict is None:
```

Comment:
But for `init_lora_eva_weights` the dataloader can actually be `None` if state dict is provided. I added a `None` check to `_get_eva_state_dict` where the dataloader is actually needed

Reply:
> But for init_lora_eva_weights the dataloader can actually be None if state dict is provided.

Well, here we already checked that `eva_state_dict is None`, so that would be covered.

> I added a `None` check to `_get_eva_state_dict` where the dataloader is actually needed

That's also fine, then the type annotation should be updated to `dataloader: Optional[Iterable]`

Reply:
sorry, which type annotation do you mean, I saw its not updated in the docstring of `init_lora_eva_weights` and added it there.

In `get_eva_state_dict` and `_get_eva_state_dict` its not really optional as it will always raise an error.

Reply:
IMO it is optional because the function accepts and expects that `None` could be passed. I would expect a type checker to complain otherwise, as the `get_eva_state_dict` call here passes `dataloader`, which according to this function could be `None`.

It's no big deal overall, I personally don't use type checkers on this code base, but I know that some users complain sometimes because their IDE warns about the wrong type being used.

Reply:
Okay that makes sense, I moved to `None` check to `init_lora_eva_weights` which was your original suggestion anyway. So in the other functions the `dataloader` argument would not be optional anymore now.